File: Algorithms/Divide and Conquer/merge sort.py
/*
   ProblemStatement: Merge Sort in Python using divide and conquer algorithm
   MergeSort is best example of a divide-and-conquer algorithm. 
   The algorithm divides the array in two halves, recursively sorts them and 
   finally merges the two sorted halves.
 */
 
 def merge(low,mid,high):
    global roll
    temp = []
    h =  low
    j = mid + 1
    while((h <= mid) and  (j<=high)):
        if roll[h] < roll[j]:
            temp.append(roll[h])
            h = h + 1
        else:
            temp.append(roll[j])
            j = j + 1
    if h > mid :
        for k in range(j,high+1):
            temp.append(roll[k])
    if j > high :
        for k in range(h,mid+1):
            temp.append(roll[k])
    temp.reverse()
    for k in range(low,high+1):
        roll[k] = temp.pop()
        
 def merge_sort(low,high):
    if low < high:
        mid = (low + high) // 2
        merge_sort(low,mid)
        merge_sort(mid+1,high)
        merge(low,mid,high)  
        
 roll = [41, 5, 82, 9, 12, 45, 49, 32, 96] /*roll is the list of items to be sorted */
 merge_sort(0, len(roll)-1)
 print("Sorted roll-",roll)

File: MergeSort.c
#include<stdio.h>

void merge(int a[11],int low, int mid, int high)
{
	int l1,l2,i,b[11];
	for(l1=low,l2=mid+1,i=low;l1<=mid&&l2<=high;i++)
	{
		if(a[l1]<=a[l2])
			b[i] = a[l1++];
		else
			b[i] = a[l2++];
	}
	while (l1<=mid)
		b[i++] = a[l1++];
	while(l2<=high)
		b[i++] = a[l2++];
	for(i=low;i<=high;i++)
		a[i]=b[i];
}
void sort(int a[], int low, int high)
{ 	
	int mid;
	if(low<high)
	{
		mid=(low+high)/2;
		sort(a,low,mid);
		sort(a,mid+1,high);
		merge(a,low,mid,high);
	}

}
int main()
{
	int i,a[11],n;
	printf("Enter number of elements:");
	scanf("%d",&n);
	printf("Enter elements");
	for(i=0;i<n;i++)
		scanf("%d",&a[i]);
	sort(a,0,n-1);
	printf("List after sorting:");
	for(i=0;i<n;i++)
		printf("%d ",a[i]);
	return 0;	
}
